### ISSUE TITLE
Support changing the thread pool size during runtime

### ISSUE THREAD

**Issue by Hoeze** [pola-rs/polars#6157]
### Problem description

I would like to use polars lazy loading inside a pypsark UDF.
Therefore, it is necessary that polars runs in single-threaded mode to avoid thread explosions.
Most other Python libraries can be limited to single-thread execution using `threadpoolctl.threadpool_limits`.
However, in polars this has no effect and it keeps using all available resources:
```python
import polars as pl
import pandas as pd
import numpy as np

df = pd.DataFrame(
    {
        "A": [1, 2, 3, 4, 5],
        "fruits": ["banana", "banana", "apple", "apple", "banana"],
        "B": [5, 4, 3, 2, 1],
        "cars": ["beetle", "audi", "beetle", "beetle", "beetle"],
    }
)
df = pd.concat([df] * 1000)
df = pd.concat([df] * 1000)
df = pd.concat([df] * 10)
df["fruits"] = np.random.randint(100000, size=df.shape[0])
df = pl.from_pandas(df)


# Using all threads is expected here
df.sort(["fruits", "B"]).select(
    [
        "fruits",
        "cars",
        pl.lit("fruits").alias("literal_string_fruits"),
        pl.col("B").filter(pl.col("cars") == "beetle").sum(),
        pl.col("A").filter(pl.col("B") > 2).sum().over("cars").alias("sum_A_by_cars"),
        pl.col("A").sum().over("fruits").alias("sum_A_by_fruits"),
        pl.col("A").reverse().over("fruits").alias("rev_A_by_fruits"),
        pl.col("A").sort_by("B").over("fruits").alias("sort_A_by_B_by_fruits"),
    ]
)


# Threadpool limits don't work; polars is still using all threads
from threadpoolctl import threadpool_limits
with threadpool_limits(limits=1):
    df.sort(["fruits", "B"]).select(
        [
            "fruits",
            "cars",
            pl.lit("fruits").alias("literal_string_fruits"),
            pl.col("B").filter(pl.col("cars") == "beetle").sum(),
            pl.col("A").filter(pl.col("B") > 2).sum().over("cars").alias("sum_A_by_cars"),
            pl.col("A").sum().over("fruits").alias("sum_A_by_fruits"),
            pl.col("A").reverse().over("fruits").alias("rev_A_by_fruits"),
            pl.col("A").sort_by("B").over("fruits").alias("sort_A_by_B_by_fruits"),
        ]
    )
```

Would it be possible to support threadpool_limits @ritchie46?
(Maybe related to #4558)

**Comment by thomasfrederikhoeck** [comment]
Would using the env var solve it for you: https://stackoverflow.com/questions/71179317/how-to-limit-number-of-threads-in-polars ?

**Comment by Hoeze** [comment]
Is there no way to specify the thread pool used by polars?

**Comment by ritchie46** [comment]
Our thread pool is not behind a lock, so we cannot modify it once set.

**Comment by ghuls** [comment]
You can set the threads before you start pyspark: 
```
export POLARS_MAX_THREADS=2
python test.py
```

**Comment by hanslovsky** [comment]
I think a dynamic thread pool would be a great enhancement. It is not crucial for me but certainly a nice to have.  Unfortunately, I don't have any informed suggestions for the implementation because my rust knowledge is rudimentary at best. Here are some ideas that may or may not be useful. Please be blunt in rejecting them if they do not sound reasonable. They all require that there is a non-global thread pool (the global thread pool can still be used for fallback):
 - Pass an optional thread pool to `DataFrame` at the time of construction. Fall back to the global thread pool defined by `POLARS_MAX_THREADS` if not provided
 - Pass an optional thread pool to calls of `compute`. Fall back to the global thread pool defined by `POLARS_MAX_THREADS` if not provided

I don't know how feasible these solutions would be. My intuition is that it would require a lot of changes to the code because now the optional thread pool would need to be passed along pretty much everywhere.

I don't know how the effort would compare against `arc_swap`.

**Comment by lorentzenchr** [comment]
Is there any chance this feature could be implemented or is https://github.com/pola-rs/polars/issues/6157#issuecomment-1377420903 basically a categorical no?

There are several real use cases where limiting the number of threads is crucial to not end up in over-subscription which hits runtime performance very hard.

**Comment by alonme** [comment]
@ritchie46 Are you open for a solution for this?

you suggested this in the past?
>Our thread pool is not behind a lock, so we cannot modify it once set. Maybe we can explore putting it behind https://docs.rs/arc-swap/0.4.7/arc_swap/.

**Comment by adamreeve** [comment]
I just want to mention another use case for controlling the thread pool size, which is to prevent numerical differences in results between different runtime environments. For example, `pl.sum_horizontal` can give different results depending on the size of the thread pool.

This can be controlled for by using the `POLARS_MAX_THREADS` environment variable, but that isn't always feasible. For example the Polars code might be used in an application outside of  your control, and if that application uses Polars elsewhere the thread pool might already be created before you get a chance to set that environment variable.

**Comment by ritchie46** [comment]
> For example, pl.sum_horizontal with floating point data can give different results depending on the size of the thread pool.

I think we should solve that.

**Comment by orlp** [comment]
@adamreeve 

> I just want to mention another use case for controlling the thread pool size, which is to prevent numerical differences in results between different runtime environments. For example, pl.sum_horizontal with floating point data can give different results depending on the size of the thread pool.

Polars doesn't offer deterministic floating point operations in general, as it would be too restrictive in parallelism. If you need exact numerics, use decimals.

**Comment by orlp** [comment]
After a little investigation I think we are open to doing this by putting the thread pool behind `arc_swap`, but first I want to centralize our Polars config and add `pl.reload_config()` (name not yet finalized) which will re-read all env vars to update our config. Then during this function we could change the thread pool as well.

PR's at this time not welcome, I'd like to do this properly myself as it involves quite a bit of core architecture but currently the streaming engine has higher priority tasks.

**Comment by adamreeve** [comment]
> Polars doesn't offer deterministic floating point operations in general, as it would be too restrictive in parallelism. If you need exact numerics, use decimals.

We don't need exact results, but we do need results to be reproducible. Does this mean that the same query could give different results on the same machine?

**Comment by orlp** [comment]
> Does this mean that the same query could give different results on the same machine?

Yes. In general floating point arithmetic is non-deterministic in Polars.

Floating point addition is not associative which means that `(a + b) + c` does not always give the same result as `a + (b + c)`. And all kinds of effects could result in for example a sum being split up differently among threads or partitions, etc, which thus necessarily affects the final result.

**Comment by adamreeve** [comment]
> For example, floating point addition is not associative which means that (a + b) + c does not always give the same result as a + (b + c). And all kinds of effects could result in for example a sum being split up differently among threads or partitions, etc, which thus necessarily affects the final result.

This is true but plenty of numerical computation libraries do offer ways to ensure deterministic results. Eg. we use MKL which has some environment variables that can be set to configure this: https://www.intel.com/content/www/us/en/docs/onemkl/developer-guide-linux/2023-0/obtaining-numerically-reproducible-results.html

But I understand it would be a large undertaking to provide this functionality, and it's off topic for this issue, so I'll leave it at that.

**Comment by adamreeve** [comment]
> in some circumstances degrade by more than a half

This is referring to not taking code paths that use AVX2. I doubt you'd see that kind of slow down by using deterministic partitioning. A couple of sentence later is this:

> Even if the code branch is not restricted, performance can degrade by 10-20% because the new functionality restricts algorithms to maintain the order of operations.

**Comment by orlp** [comment]
@adamreeve We'd have *both* restrictions if we'd require determinism. And more that Intel's library doesn't have to worry about, e.g. the order of morsels in the streaming engine, etc.

**Comment by lorentzenchr** [comment]
If I may, I'd like put the focus of this issue (back) to the control of threads in polars.

I would suggest to have the discussion about floating point arithmetic ([IEEE_754](https://en.wikipedia.org/wiki/IEEE_754)) elsewhere. My experience tells me that it will be fruitless, the last few bits of floating point numbers just have too many influencers in a system (from compiler to selected instructions during and order of execution), see e.g. https://doi.org/10.3929/ethz-b-000304330.

@orlp As to thread control, anything you could use help with?

**Comment by kdekker-kdr4** [comment]
Any update on this issue @lorentzenchr ?
